src/psyclone/psyir/transformations/matmul2code_trans.py

```diff
@@ -33,20 +33,64 @@
-MATMUL operator to PSyIR code. This could be useful if the
-MATMULoperator is not supported by the back-end or if the performance
-in the inline code is better than the intrinsic. MATMUL supports both
-matrix multiply and matrix vector multiply. At the moment this
-transformation is limited to matrix vector multiply.
+'''Module providing a transformation from a PSyIR MATMUL operator to
+PSyIR code. This could be useful if the MATMUL operator is not
+supported by the back-end or if the performance in the inline code is
+better than the intrinsic. MATMUL supports both matrix multiply and
+matrix vector multiply. At the moment this transformation is limited
+to matrix vector multiply.
 
 '''
 from __future__ import absolute_import
 from psyclone.psyir.nodes import BinaryOperation, Assignment, Reference, \
     Loop, Literal, Array, Range
-from psyclone.psyir.symbols import DataType, DataSymbol
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE, REAL_TYPE, \
+    ArrayType
 from psyclone.psyGen import Transformation
-from psyclone.errors import GenerationError
+
+
+def _get_array_bound(array, index):
+    '''A utility function that returns the appropriate loop bounds (lower,
+    upper and step) for a particular index of an array reference. At
+    the moment all entries for the index are assumed to be accessed.
+    If the array symbol is declared with known bounds (an integer or a
+    symbol) then these bound values are used. If the size is unknown
+    (a deferred or attribute type) then the LBOUND and UBOUND PSyIR
+    nodes are used.
+
+    :param array: the reference that we are interested in.
+    :type array: :py:class:`psyir.nodes.Reference`
+    :param int index: the (array) reference index that we are \
+        interested in.
+    :returns: the loop bounds for this array index.
+    :rtype: (Literal, Literal, Literal) or \
+        (BinaryOperation, BinaryOperation, Literal)
+
+    :raises TransformationError: if the shape of the array's symbol is \
+        not supported.
+
+    '''
+    my_dim = array.symbol.shape[index]
+    if isinstance(my_dim, int):
+        lower_bound = Literal("1", INTEGER_TYPE)
+        upper_bound =  Literal(str(my_dim), INTEGER_TYPE)
+    elif isinstance(my_dim, DataSymbol):
+        lower_bound = Literal("1", INTEGER_TYPE)
+        upper_bound =  Reference(my_dim)
+    elif my_dim in [ArrayType.Extent.DEFERRED, ArrayType.Extent.ATTRIBUTE]:
+        lower_bound = BinaryOperation.create(
+            BinaryOperation.Operator.LBOUND, Reference(array.symbol),
+            Literal(str(index), INTEGER_TYPE))
+        upper_bound =  BinaryOperation.create(
+            BinaryOperation.Operator.UBOUND, Reference(array.symbol),
+            Literal(str(index), INTEGER_TYPE))
+    else:
+        # Added import here to avoid circular dependencies.
+        from psyclone.psyir.transformations import TransformationError
+        raise TransformationError("Unsupported index type '{0}' found for array '{1}'."
```

Comment:
Please also include which index of the array is causing the problem.

Reply:
Added. I called it "dimension" rather than "index" as that made more sense to me. Hopefully that is OK.